sympy/assumptions/handlers/sets.py

```diff
@@ -31,682 +38,753 @@ def _number(expr, assumptions):
-    def Pow(expr, assumptions):
-        return expr.exp.is_Rational and ask(
-            Q.algebraic(expr.base), assumptions)
-
-    @staticmethod
-    def Rational(expr, assumptions):
-        return expr.q != 0
-
-    Float, GoldenRatio, TribonacciConstant, ImaginaryUnit, AlgebraicNumber = \
-        [staticmethod(CommonHandler.AlwaysTrue)]*5
-
-    Infinity, NegativeInfinity, ComplexInfinity, Pi, Exp1 = \
-        [staticmethod(CommonHandler.AlwaysFalse)]*5
-
-    @staticmethod
-    def exp(expr, assumptions):
-        x = expr.args[0]
-        if ask(Q.algebraic(x), assumptions):
-            return ask(~Q.nonzero(x), assumptions)
-
-    @staticmethod
-    def cot(expr, assumptions):
-        x = expr.args[0]
-        if ask(Q.algebraic(x), assumptions):
+@AntiHermitianHandler.register(Pow)
+def _(expr, assumptions):
+    """
+    Hermitian**Integer  -> !Antihermitian
+    Antihermitian**Even -> !Antihermitian
+    Antihermitian**Odd  -> Antihermitian
+    """
+    if expr.is_number:
+        return None
+    if ask(Q.hermitian(expr.base), assumptions):
+        if ask(Q.integer(expr.exp), assumptions):
             return False
+    elif ask(Q.antihermitian(expr.base), assumptions):
+        if ask(Q.even(expr.exp), assumptions):
+            return False
+        elif ask(Q.odd(expr.exp), assumptions):
+            return True
 
-    @staticmethod
-    def log(expr, assumptions):
-        x = expr.args[0]
-        if ask(Q.algebraic(x), assumptions):
-            return ask(~Q.nonzero(x - 1), assumptions)
+@AntiHermitianHandler.register(MatrixBase)
+def _(mat, assumptions):
+    rows, cols = mat.shape
+    ret_val = True
+    for i in range(rows):
+        for j in range(i, cols):
+            cond = fuzzy_bool(Eq(mat[i, j], -conjugate(mat[j, i])))
+            if cond == None:
+                ret_val = None
+            if cond == False:
+                return False
+    return ret_val
+
+
+# AlgebraicHandler
+
+AlgebraicHandler = Dispatcher(
+    "AlgebraicHandler",
+    doc="""Handler for Q.algebraic key. """
+)
+
+for cls in (AlgebraicNumber, Float, GoldenRatio, ImaginaryUnit,
+            TribonacciConstant):
+    AlgebraicHandler.register(cls)(lambda expr, assumptions: True)
+
+for cls in (ComplexInfinity, Exp1, Infinity, NaN, NegativeInfinity, Pi):
+    AlgebraicHandler.register(cls)(lambda expr, assumptions: False)
+
+@AlgebraicHandler.register(Add)
+@AlgebraicHandler.register(Mul)
+def _(expr, assumptions):
+    return test_closed_group(expr, assumptions, Q.algebraic)
+
+@AlgebraicHandler.register(Pow)
+def _(expr, assumptions):
+    return expr.exp.is_Rational and ask(Q.algebraic(expr.base), assumptions)
+
+@AlgebraicHandler.register(Rational)
+def _(expr, assumptions):
+    return expr.q != 0
+
+@AlgebraicHandler.register(asin)
+@AlgebraicHandler.register(atan)
+@AlgebraicHandler.register(cos)
+@AlgebraicHandler.register(exp)
+@AlgebraicHandler.register(sin)
+@AlgebraicHandler.register(tan)
+def _(expr, assumptions):
```

Comment:
Perhaps it could be made possible to pass a list to `register` or perhaps there could be a `register_many`

Reply:
Good idea. Or maybe it would be better to use loop like
```python
for cls in (asin, atan, ...):
    AlgebraicHandler.register(func)
```

Reply:
There are lots of loops already. I prefer the decorator style but it could be made more compact for cases like this.

Reply:
I implemented `register_many` method.